programs/stake/src/stake_state.rs

```diff
@@ -64,489 +57,90 @@ pub(crate) fn null_tracer() -> Option<impl FnMut(&InflationPointCalculationEvent
-        Self::from(account).and_then(|state: Self| state.meta())
-    }
-
-    pub fn meta(&self) -> Option<Meta> {
-        match self {
-            StakeState::Stake(meta, _stake) => Some(*meta),
-            StakeState::Initialized(meta) => Some(*meta),
-            _ => None,
-        }
-    }
-}
-
-#[derive(Debug, Serialize, Deserialize, PartialEq, Clone, Copy, AbiExample)]
-pub enum StakeAuthorize {
-    Staker,
-    Withdrawer,
-}
-
-#[derive(Default, Debug, Serialize, Deserialize, PartialEq, Clone, Copy, AbiExample)]
-pub struct Lockup {
-    /// UnixTimestamp at which this stake will allow withdrawal, unless the
-    ///   transaction is signed by the custodian
-    pub unix_timestamp: UnixTimestamp,
-    /// epoch height at which this stake will allow withdrawal, unless the
-    ///   transaction is signed by the custodian
-    pub epoch: Epoch,
-    /// custodian signature on a transaction exempts the operation from
-    ///  lockup constraints
-    pub custodian: Pubkey,
-}
-
-impl Lockup {
-    pub fn is_in_force(&self, clock: &Clock, custodian: Option<&Pubkey>) -> bool {
-        if custodian == Some(&self.custodian) {
-            return false;
-        }
-        self.unix_timestamp > clock.unix_timestamp || self.epoch > clock.epoch
+        Self::from(account).and_then(|state: StakeState| state.meta())
     }
-}
-
-#[derive(Default, Debug, Serialize, Deserialize, PartialEq, Clone, Copy, AbiExample)]
-pub struct Authorized {
-    pub staker: Pubkey,
-    pub withdrawer: Pubkey,
-}
-
-#[derive(Default, Debug, Serialize, Deserialize, PartialEq, Clone, Copy, AbiExample)]
-pub struct Meta {
-    pub rent_exempt_reserve: u64,
-    pub authorized: Authorized,
-    pub lockup: Lockup,
-}
 
-impl Meta {
-    pub fn set_lockup(
-        &mut self,
-        lockup: &LockupArgs,
-        signers: &HashSet<Pubkey>,
-        clock: Option<&Clock>,
-    ) -> Result<(), InstructionError> {
-        match clock {
-            None => {
-                // pre-stake_program_v4 behavior: custodian can set lockups at any time
-                if !signers.contains(&self.lockup.custodian) {
-                    return Err(InstructionError::MissingRequiredSignature);
-                }
-            }
-            Some(clock) => {
-                // post-stake_program_v4 behavior:
-                // * custodian can update the lockup while in force
-                // * withdraw authority can set a new lockup
-                //
-                if self.lockup.is_in_force(clock, None) {
-                    if !signers.contains(&self.lockup.custodian) {
-                        return Err(InstructionError::MissingRequiredSignature);
-                    }
-                } else if !signers.contains(&self.authorized.withdrawer) {
-                    return Err(InstructionError::MissingRequiredSignature);
-                }
+    // TODO move VoteState into solana_program to move this function into Stake
```

Comment:
What criteria do you think we should use to decide if/when to do this move?

Reply:
That's a good question.  Is there any desire to also move the vote program types into `solana_program`?  From my limited knowledge of the code, I think it'll be really tricky.  If there's no desire to move it, I can git rid of this `TODO` and instead say, `This can be moved into solana_program if VoteState is ever also included in `solana_program`

Reply:
> I can git rid of this `TODO`

This sg to me :)

Reply:
Done in 0d298a9